### DISCUSSION TITLE
Documentation please

### DISCUSSION THREAD

**Discussion by madnexus** [le0pard/certonid#7] [Q&A]
Is there any updated version of the docs? Tried to get it to work but it was impossible due to the confusing documentation.

**Comment by le0pard** [comment]
@madnexus what exactly doesn't work from wiki?

P.S. documentation is up to date in wiki.

**Comment by madnexus** [comment]
Hi @le0pard and thanks for the reply. 

There is a few things I would like to ask/mention:

- What is exactly the kmsauth.service_id? In some parts of the docs is described as "KMS Auth service ID for encryption context" while in other like just name_for_service. I'm still not too sure about what does it need to be set here.
- Why there's a reference about IAM users for account administration and access control to the certonid function in the mailtrap blog but no mention whatsoever in the certonid wiki?
- The KMS authentication section in the wiki is a bit confusing. The default config example includes the kmsauth section which is independent and easily confused with the ca key and passphrase encryption capabilities of Certonid (and also way before that you read the KMS Authentication section which is the last page of the docs).
- After going around the setup and permissions a few times and discarding the KMS Auth for now, I'm getting the following error:

Not using KMS auth in the certonid.yml I'm getting a "Error to invoke AWS Lambda function: AccessDeniedException: status code: 403"

No luck so far so wondering what could I have missed! 

Thank you

**Comment by le0pard** [comment]
> What is exactly the kmsauth.service_id? In some parts of the docs is described as "KMS Auth service ID for encryption context" while in other like just name_for_service. I'm still not too sure about what does it need to be set here.

`service_id` is key, which used in KMSAuth. Inside encryption context it used in key `to`. This key need to be known in serverless and cli components. More info how works KMSAuth you can get here: https://lyft.github.io/confidant/advanced/kms_auth/

In docs clearly shown that `name_for_service` is value for `kmsauth.service_id` only on https://github.com/certonid/certonid/wiki/KMS-authentication page.

> Why there's a reference about IAM users for account administration and access control to the certonid function in the mailtrap blog but no mention whatsoever in the certonid wiki?

Yep, this need to be added. You just need setup aws user for certonid cli to allow to involke serverless componnet by AWS credentials.

Added example of AWS role in doc: https://github.com/certonid/certonid/wiki/Certonid-CLI-Config

![image](https://user-images.githubusercontent.com/98444/88464406-2735c000-cec3-11ea-8416-9d98f97a864e.png)


> The KMS authentication section in the wiki is a bit confusing. The default config example includes the kmsauth section which is independent and easily confused with the ca key and passphrase encryption capabilities of Certonid (and also way before that you read the KMS Authentication section which is the last page of the docs).

As written in doc:
"After this you need to modify certonid.yml config for serverless", which mean this part added in **serverless component** and position in yml correct. Here you can see example for **serverless component**: "

https://github.com/certonid/certonid/wiki/Serverless-Config , where kmsauth is just separate config part (because it is optional).

In CLI kmsauth yml part should be part as SSH config for named key: https://github.com/certonid/certonid/wiki/Certonid-CLI-Config

This is on last wiki page, because KMSAuth is optional, not required. It is only needed if you additionally you need check `username`, which added in key from CLI request.

> After going around the setup and permissions a few times and discarding the KMS Auth for now, I'm getting the following error:
Not using KMS auth in the certonid.yml I'm getting a "Error to invoke AWS Lambda function: AccessDeniedException: status code: 403"

This mean, that AWS user, which used to call your serverless component from CLI is not have access to call AWS lambda function.

Hope this answer on your questions.

**Comment by madnexus** [comment]
Hi @le0pard 

Sorry for the late reply! 

Found the main issue was sometimes applying policies and perms takes a while in AWS. Sometimes up to 4 or 5 minutes! Due to this delay we binned some tests and policies thinking they weren't working. 

Will mention a few things soon as we have some questions.

Thanks again for the quick replies and the effort of developing this! :)

**Comment by le0pard** [comment]
@madnexus thanks for feedback

**Comment by le0pard** [comment]
Any updates @madnexus ?
